File: src/utils/daily-game-storage.ts
import { DailyGameEntry, DailyGameHistory } from '../types/daily-menu';

const STORAGE_KEY = 'gamedeck_daily_history';

export const getDailyGameHistory = (): DailyGameHistory => {
  try {
    const stored = localStorage.getItem(STORAGE_KEY);
    return stored ? JSON.parse(stored) : {};
  } catch (error) {
    console.error('Error reading daily game history:', error);
    return {};
  }
};

export const getTodayDate = (): string => {
  return new Date().toISOString().split('T')[0];
};

export const getTodayGames = (): DailyGameEntry[] => {
  const history = getDailyGameHistory();
  const today = getTodayDate();
  return history[today] || [];
};

export const addGameToToday = (gameId: number, gameName: string): void => {
  const history = getDailyGameHistory();
  const today = getTodayDate();
  const todayGames = history[today] || [];

  // Check if game already played today
  if (todayGames.some(entry => entry.gameId === gameId)) {
    return;
  }

  const newEntry: DailyGameEntry = {
    gameId,
    gameName,
    playedDate: new Date().toISOString(),
    completed: false,
  };

  history[today] = [...todayGames, newEntry];
  localStorage.setItem(STORAGE_KEY, JSON.stringify(history));
};

export const markGameAsCompleted = (gameId: number): void => {
  const history = getDailyGameHistory();
  const today = getTodayDate();
  const todayGames = history[today] || [];

  const updatedGames = todayGames.map(entry =>
    entry.gameId === gameId ? { ...entry, completed: true } : entry
  );

  history[today] = updatedGames;
  localStorage.setItem(STORAGE_KEY, JSON.stringify(history));
};

export const getPlayedGamesCount = (): number => {
  const history = getDailyGameHistory();
  return Object.values(history).reduce((total, dayGames) => total + dayGames.length, 0);
};

export const getStreakDays = (): number => {
  const history = getDailyGameHistory();
  const dates = Object.keys(history).sort().reverse();

  let streak = 0;
  let currentDate = new Date();

  for (const dateStr of dates) {
    const checkDate = currentDate.toISOString().split('T')[0];
    if (dateStr === checkDate) {
      streak++;
      currentDate.setDate(currentDate.getDate() - 1);
    } else {
      break;
    }
  }

  return streak;
};

File: ios-app/GameDeck/Services/DailyGameStorage.swift
import Foundation

class DailyGameStorage {
    static let shared = DailyGameStorage()
    private let storageKey = "gamedeck_daily_history"

    private init() {}

    // MARK: - Storage Operations

    func getDailyGameHistory() -> [String: [DailyGameEntry]] {
        guard let data = UserDefaults.standard.data(forKey: storageKey),
              let history = try? JSONDecoder().decode([String: [DailyGameEntry]].self, from: data) else {
            return [:]
        }
        return history
    }

    func getTodayDate() -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter.string(from: Date())
    }

    func getTodayGames() -> [DailyGameEntry] {
        let history = getDailyGameHistory()
        let today = getTodayDate()
        return history[today] ?? []
    }

    func addGameToToday(gameId: Int, gameName: String) {
        var history = getDailyGameHistory()
        let today = getTodayDate()
        var todayGames = history[today] ?? []

        // Check if game already played today
        if todayGames.contains(where: { $0.gameId == gameId }) {
            return
        }

        let newEntry = DailyGameEntry(
            gameId: gameId,
            gameName: gameName,
            playedDate: Date(),
            completed: false
        )

        todayGames.append(newEntry)
        history[today] = todayGames
        saveHistory(history)
    }

    func markGameAsCompleted(gameId: Int) {
        var history = getDailyGameHistory()
        let today = getTodayDate()
        guard var todayGames = history[today] else { return }

        if let index = todayGames.firstIndex(where: { $0.gameId == gameId }) {
            todayGames[index].completed = true
            history[today] = todayGames
            saveHistory(history)
        }
    }

    func getPlayedGamesCount() -> Int {
        let history = getDailyGameHistory()
        return history.values.reduce(0) { $0 + $1.count }
    }

    func getStreakDays() -> Int {
        let history = getDailyGameHistory()
        let sortedDates = history.keys.sorted(by: >)

        var streak = 0
        var currentDate = Date()
        let calendar = Calendar.current
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"

        for dateStr in sortedDates {
            let checkDate = formatter.string(from: currentDate)
            if dateStr == checkDate {
                streak += 1
                currentDate = calendar.date(byAdding: .day, value: -1, to: currentDate) ?? currentDate
            } else {
                break
            }
        }

        return streak
    }

    private func saveHistory(_ history: [String: [DailyGameEntry]]) {
        if let encoded = try? JSONEncoder().encode(history) {
            UserDefaults.standard.set(encoded, forKey: storageKey)
        }
    }
}
